unix,win: add uv_get_usec_since_epoch

refack opened the pull request:
~Change signature to `int64_t uv_gettimeofday()`, based on https://github.com/libuv/libuv/issues/2243~

~I thought about changing the name to `uv_get_usec_since_epoch` but that might be too much...~

~CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1335/~

Went with a new `int64_t uv_get_usec_since_epoch(void)` after all.

bnoordhuis commented on docs/src/misc.rst:
```diff
@@ -579,9 +579,9 @@ API
     .. versionadded:: 1.25.0
 
-.. c:function:: int uv_gettimeofday(uv_timeval_t* tv)
+.. c:function:: int64_t uv_gettimeofday()
```
`uv_gettimeofday(void)` (compare `uv_hrtime(void)`.)

refack replied:
```suggestion
.. c:function:: int64_t uv_gettimeofday(void)
```

bnoordhuis commented on docs/src/misc.rst:
```diff
@@ -579,9 +579,9 @@ API
-    Cross-platform implementation of :man:`gettimeofday(2)`. The timezone
-    argument to `gettimeofday()` is not supported, as it is considered obsolete.
+    Cross-platform implementation of micro-seconds since epoch. Similar to
+    :man:`gettimeofday(2)`. Negative returned values are error codes.
```
So, on many systems it's technically possible to set the system clock to a date before January 1, 1970; that's why `tv_sec` is signed, to support such dates.

I don't know if it's actually useful to support that but in that case libuv should detect such dates and return an error, not just a semi-random value < 0.

I see that you have an `assert(ret > 0)` now. I guess that's okay?

refack replied:
IIUC supporting `tv_sec < 0` is mutually exclusive with Y2K38...
I think we should pick a side (either cast to unsigned or bail).

bnoordhuis replied:
This function overlaps too much with `uv_gettimeofday()`, IMO. I'd rather remove it.

> IIUC supporting tv_sec < 0 is mutually exclusive with Y2K38...

Not sure I follow. Not if `tv_sec` is 64 bits.

cjihrig replied:
I'd rather remove this function to. Or, at the very least, open a separate PR for it.

I'd like to get the `uv_timeval64_t` changes landed before the next release.

refack replied:
I'll spin this off to a new PR.

bnoordhuis commented on src/win/util.c:
```diff
@@ -1779,19 +1779,20 @@ int uv_os_uname(uv_utsname_t* buffer) {
+  ret = file_time.dwHighDateTime;
+  assert(file_time.dwHighDateTime < LONG_MAX);
+  ret <<= sizeof(file_time.dwHighDateTime) * CHAR_BIT;  /* 4 * 8 = 32 */
+  ret += file_time.dwHighDateTime;
```
`ret += file_time.dwLowDateTime`?

refack replied:
```suggestion
  ret += file_time.dwHighDateTime;
```
Ack (I keep thinking of a better way to test my code...)

cjihrig commented:
FWIW, this is not what I had in mind in #2243. I assumed we would define a new data type similar to `uv_timeval_t`, but with `int64_t`s. Then, in v2, the old `uv_timeval_t` could go away.

refack commented:
> FWIW, this is not what I had in mind

I guess that's me thinking in WIN32 terms... We could have two implementations, since this one is more of a `uv_get_usec_since_epoch` then a cross platform `gettimeofday`.

cjihrig commented:
@refack do you plan to implement https://github.com/libuv/libuv/pull/2244#issuecomment-478659463? If not, I can take a shot at it.

refack commented:
> @refack do you plan to implement [#2244 (comment)](https://github.com/libuv/libuv/pull/2244#issuecomment-478659463)? If not, I can take a shot at it.

I'll do that.

refack commented:
* Added `uv_timeval64_t`
* changed uv_gettimeofday to 
  ```c
  int intuv_gettimeofday(&uv_timeval64_t tv)
  ```
* added
  ```c
  int64_t uv_get_usec_since_epoch();
  ```
* Documented that `uv_get_usec_since_epoch` prefers Y2K38 over support for pre-epoch dates

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1336/

saghul commented on docs/src/misc.rst:
```diff
@@ -579,9 +579,18 @@ API
 
     Cross-platform implementation of :man:`gettimeofday(2)`. The timezone
     argument to `gettimeofday()` is not supported, as it is considered obsolete.
+	In order to support dates after Y2K38 this does not support dates before
```
style: indentation seems off here

saghul commented on src/unix/core.c:
```diff
@@ -1444,3 +1444,17 @@ int uv_gettimeofday(uv_timeval_t* tv) {
+  if (gettimeofday(&time, NULL) != 0)
+    return UV__ERR(errno);
+
+	/* cast to unsigned to prefer Y2K38 over dates < 1970 */
```
style: indentation seems off here

bnoordhuis commented on include/uv.h:
```diff
@@ -1099,6 +1099,11 @@ typedef struct {
 
+typedef struct {
+  int64_t tv_sec;
+  int64_t tv_usec;
```
`tv_usec` could be `uint32_t`. There generally aren't more than one million microseconds in a second. :-)

It might make sense to upgrade it to nanoseconds.

refack replied:
Alignment ?
Anyway `nsec` it is

bnoordhuis commented on src/unix/core.c:
```diff
@@ -1444,3 +1444,17 @@ int uv_gettimeofday(uv_timeval_t* tv) {
+    return UV__ERR(errno);
+
+	/* cast to unsigned to prefer Y2K38 over dates < 1970 */
+	ret = (uint64_t)time.tv_sec * 1000000;
```
`int64_t`?

refack replied:
on "03:14:08 UTC on 19 January 2038" time.tv_sec will equal `b‭1000'0000'0000'0000'0000'0000'0000'0000‬` so we can interpret that either as and unsigned `2147483648` or a signed `-2147483648`.

refack replied:
https://coliru.stacked-crooked.com/a/1ad3777fd9b0197a

refack replied:
Yes, casting was wrong.

refack commented:
https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1339/

I've fixed the docs to make it clear that `uv_gettimeofday` supportes negative system time, but does not support Y2K38, while `uv_get_usec_since_epoch` does not support negative system time, but does support Y2K38.

santigimeno commented:
Would it make sense using [clock_gettime(CLOCK_REALTIME)](https://linux.die.net/man/2/clock_gettime) where supported?

cjihrig commented on src/unix/core.c:
```diff
@@ -1440,7 +1440,22 @@ int uv_gettimeofday(uv_timeval_t* tv) {
-  tv->tv_sec = time.tv_sec;
-  tv->tv_usec = time.tv_usec;
+  tv->tv_sec = (int64_t)time.tv_sec;
+  tv->tv_nsec = (int64_t)time.tv_usec * 1000;
```
Can we stick with `usec` to make it more of a drop in replacement, and so that users aren't confused?

refack replied:
But then it doesn't make sense to use 64bit int for a value that should be < 1'000'000

cjihrig replied:
Maybe not, but making this nanoseconds seems like a sure fire footgun for anyone using this that is familiar with `gettimeofday()`.

refack commented:
Thinking about the comments and the design issues maybe changing `uv_gettimeofday` isn't the best way to go.
1. 32/64 issue:
   * If we use two 64bit fields it's overkill as any reasonable value can fit in a total of 64bit
   * Makeing `tv_usec` 64bit is wastefull since it's value should e < 1'000'000
   * Changing `/s/tv_usec/tv_nsec` is confusing WRT to POSIX `gettimeofday`
   * Since the underlying system call returns a signed 32bit value for seconds, we need to choose pre-epoch dates or post Y2K38. Either one is opinionated.
2. Since it seems the use case for this is wall clock IMO `uv_get_usec_since_epoch` makes more sense since it does not imply similarity to `gettimeofday(2)`. Does not require an out-param, and is guaranteed not to err on WIN32 (and much much less likely to err on POSIX)...

So I just left the new function in this PR

cjihrig commented:
This has conflicts and seems to have stalled out. I'm going to close this out for now.
